File: xero/lru_cache.py
# Implement a LRU cache - Leetcode
# get
# put

from typing import Dict


class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LRUCache:
    _rightNode: Node = Node(0, 0)  # Just recent used
    _leftNode: Node = Node(0, 0)  # Last recent used
    _cache: Dict[int, Node] = {}

    def __init__(self, capacity: int):
        self._capacity = capacity
        self._rightNode.prev = self._leftNode
        self._leftNode.next = self._rightNode

    def get(self, key: int) -> int:
        if key not in self._cache:
            return -1

        # Move the node to the right of the cache
        self._remove(self._cache[key])
        self._insert(self._cache[key])

        return self._cache[key].value

    def put(self, key: int, value: int) -> None:
        if key in self._cache:
            # Update the value of the node
            self._cache[key].value = value
            self._remove(self._cache[key])
            self._insert(self._cache[key])
            return
        elif key not in self._cache:
            # Create a new node
            self._cache[key] = Node(key, value)
            self._insert(self._cache[key])

        # Remove the LRU value if the cache is full
        if len(self._cache) > self._capacity:
            self._remove(self._leftNode.next)
            self._cache.pop(self._leftNode.next.key)

    def _remove(self, node: Node) -> None:
        """Remove a node from the double-linked list"""
        left, right = node.prev, node.next
        left.next, right.prev = right, left

    def _insert(self, node: Node) -> None:
        """Insert a node on the right of the double-linked list"""
        prev = self._rightNode.prev
        prev.next, self._rightNode.prev = node, node
        node.next, node.prev = self._rightNode, prev


if __name__ == "__main__":
    lru_cache = LRUCache(2)
    lru_cache.put(1, 1)
    lru_cache.put(2, 2)
    print(lru_cache.get(1))
    lru_cache.put(3, 3)
    print(lru_cache.get(1))
    print(lru_cache.get(2))
    print(lru_cache.get(3))
    lru_cache.put(4, 4)
    print(lru_cache.get(1))
    print(lru_cache.get(3))
    print(lru_cache.get(4))

File: xero/LRUCache.cs
using System;
using System.Collections.Generic;

namespace Xero
{
    public class Node
    {
        public int Key { get; set; }
        public int Value { get; set; }
        public Node Prev { get; set; } = null!;
        public Node Next { get; set; } = null!;

        public Node(int key, int value)
        {
            Key = key;
            Value = value;
        }
    }

    public class LRUCache
    {
        private readonly int _capacity;
        private readonly Dictionary<int, Node> _cache;
        private readonly Node _leftNode; // Least Recently Used
        private readonly Node _rightNode; // Most Recently Used

        public LRUCache(int capacity)
        {
            _capacity = capacity;
            _cache = new Dictionary<int, Node>();

            // Initialize dummy nodes
            _leftNode = new Node(0, 0);
            _rightNode = new Node(0, 0);

            // Connect dummy nodes
            _leftNode.Next = _rightNode;
            _rightNode.Prev = _leftNode;
        }

        public int Get(int key)
        {
            if (!_cache.ContainsKey(key))
            {
                return -1;
            }

            // Move the node to the right of the cache
            Node node = _cache[key];
            Remove(node);
            Insert(node);

            return node.Value;
        }

        public void Put(int key, int value)
        {
            if (_cache.ContainsKey(key))
            {
                // Update the value of the node
                _cache[key].Value = value;
                Remove(_cache[key]);
                Insert(_cache[key]);
                return;
            }

            // Create a new node
            _cache[key] = new Node(key, value);
            Insert(_cache[key]);

            // Remove the LRU value if the cache is full
            if (_cache.Count > _capacity)
            {
                Node lruNode = _leftNode.Next;
                Remove(lruNode);
                _cache.Remove(lruNode.Key);
            }
        }

        private void Remove(Node node)
        {
            // Remove a node from the double-linked list
            Node left = node.Prev;
            Node right = node.Next;
            left.Next = right;
            right.Prev = left;
        }

        private void Insert(Node node)
        {
            // Insert a node on the right of the double-linked list
            Node prev = _rightNode.Prev;
            prev.Next = node;
            _rightNode.Prev = node;
            node.Next = _rightNode;
            node.Prev = prev;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Testing LRUCache");
            LRUCache lruCache = new LRUCache(2);
            lruCache.Put(1, 1);
            lruCache.Put(2, 2);
            Console.WriteLine(lruCache.Get(1));
            lruCache.Put(3, 3);
            Console.WriteLine(lruCache.Get(1));
            Console.WriteLine(lruCache.Get(2));
            Console.WriteLine(lruCache.Get(3));
            lruCache.Put(4, 4);
            Console.WriteLine(lruCache.Get(1));
            Console.WriteLine(lruCache.Get(3));
            Console.WriteLine(lruCache.Get(4));
        }
    }
}
